src/NLog/Targets/TargetWithContext.cs

```diff
@@ -0,0 +1,747 @@
+                else if (Layout is JsonLayout jsonLayout)
+                    jsonLayout.IncludeMdc = value;
+            }
+        }
+        private bool _includeMdc;
+
+        /// <inheritdoc/>
+        public bool IncludeNdc
+        {
+            get { return _includeNdc && !_contextLayout; }
+            set
+            {
+                _includeNdc = value;
+                if (Layout is IIncludeContext contextLayout)
+                    contextLayout.IncludeNdc = value;
+            }
+        }
+        private bool _includeNdc;
+
+#if !SILVERLIGHT
+        /// <inheritdoc/>
+        public bool IncludeMdlc
+        {
+            get { return _includeMdlc && !_contextLayout; }
+            set
+            {
+                _includeMdlc = value;
+                if (Layout is IIncludeContext contextLayout)
+                    contextLayout.IncludeMdlc = value;
+                else if (Layout is JsonLayout jsonLayout)
+                    jsonLayout.IncludeMdlc = value;
+            }
+        }
+        private bool _includeMdlc;
+
+        /// <inheritdoc/>
+        public bool IncludeNdlc
+        {
+            get { return _includeNdlc && !_contextLayout; }
+            set
+            {
+                _includeNdlc = value;
+                if (Layout is IIncludeContext contextLayout)
+                    contextLayout.IncludeNdlc = value;
+            }
+        }
+        private bool _includeNdlc;
+#endif
+
+        private bool _contextLayout;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        protected TargetWithContext()
+        {
+            InternalLayoutMdc = new LayoutContextMdc(this);
+            InternalLayoutNdc = new LayoutContextNdc(this);
+#if !SILVERLIGHT
+            InternalLayoutMdlc = new LayoutContextMdlc(this);
+            InternalLayoutNdlc = new LayoutContextNdlc(this);
+#endif
+            OptimizeBufferReuse = true;
+        }
+
+        /// <summary>
+        /// Initializes the target.
+        /// </summary>
+        protected override void InitializeTarget()
+        {
+            _contextLayout = false;
+            base.InitializeTarget();
+            if (Layout is IIncludeContext contextLayout)
+            {
+                if (IncludeAllProperties && !contextLayout.IncludeAllProperties)
+                    contextLayout.IncludeAllProperties = true;
+                if (IncludeMdc && !contextLayout.IncludeMdc)
+                    contextLayout.IncludeMdc = true;
+                if (IncludeNdc && !contextLayout.IncludeNdc)
+                    contextLayout.IncludeNdc = true;
+#if !SILVERLIGHT
+                if (IncludeMdlc && !contextLayout.IncludeMdlc)
+                    contextLayout.IncludeMdlc = true;
+                if (IncludeNdlc && !contextLayout.IncludeNdlc)
+                    contextLayout.IncludeNdlc = true;
+#endif
+                _contextLayout = true;
+            }
+            else if (Layout is JsonLayout jsonLayout)
```

Comment:
I'm not sure if this is good for the maintainability. Could we use/introduce an interface and handle IIncludeContext / JsonLayout  the same? (same for setters). What do you think? 

Maybe we could also check the type of Layout in init?

Reply:
I agree, and have now changed JsonLayout so it implements IIncludeContext.